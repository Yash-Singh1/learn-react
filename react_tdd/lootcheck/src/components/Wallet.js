import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import { deposit, withdraw } from '../actions/actions';

export class Wallet extends Component {
  constructor() {
    super();

    this.state = { balance: undefined };
  }

  updateBalance(event) {
    this.setState({ balance: parseInt(event.target.value, 10) });
  }

  deposit() {
    if (isNaN(parseInt(this.state.balance))) {
      return;
    }
    this.props.deposit(this.state.balance);
  }

  withdraw() {
    if (isNaN(parseInt(this.state.balance))) {
      return;
    }
    this.props.withdraw(this.state.balance);
  }

  render() {
    return (
      <div>
        <h3 className="balance">Wallet balance: {this.props.balance}</h3>
        <br />
        <input
          className="input-wallet"
          onChange={(event) => this.updateBalance(event)}
        />
        <Button className="btn-deposit" onClick={() => this.deposit()}>
          Deposit
        </Button>
        <Button className="btn-withdraw" onClick={() => this.withdraw()}>
          Withdraw
        </Button>
      </div>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return {
    balance: state.balance
  };
}

export default connect(mapStateToProps, { deposit, withdraw })(Wallet);
