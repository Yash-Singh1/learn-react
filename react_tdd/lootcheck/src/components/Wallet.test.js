import React from 'react';
import { shallow } from 'enzyme';
import { Wallet } from './Wallet';

describe('Wallet', () => {
  const mockDeposit = jest.fn();
  const mockWithrawal = jest.fn();
  const props = {
    balance: 20,
    deposit: mockDeposit,
    withdraw: mockWithrawal
  };
  const wallet = shallow(<Wallet {...props} />);

  it('renders properly', () => {
    expect(wallet.instance()).toMatchSnapshot();
  });

  it('displays the balance', () => {
    expect(wallet.find('.balance').text()).toEqual('Wallet balance: 20');
  });

  it('creates an input to deposit into or withdraw from the balance', () => {
    expect(wallet.find('.input-wallet').exists()).toBe(true);
  });

  describe('when the user types into the wallet input', () => {
    const userBalance = '25';

    beforeEach(() => {
      wallet
        .find('.input-wallet')
        .simulate('change', { target: { value: userBalance } });
    });

    it('updates the local wallet balance in `state` and converts it to a number', () => {
      expect(wallet.state().balance).toEqual(parseInt(userBalance, 10));
    });

    describe('and the user wants to make a deposit', () => {
      it('dispatches the `deposit()` it recieves from props with the local balance', () => {
        wallet.find('.btn-deposit').simulate('click');
        expect(mockDeposit).toHaveBeenCalledWith(parseInt(userBalance, 10));
      });

      it("doesn't dispatch the `deposit()` if a non number is recieved from props with local balance", () => {
        wallet
          .find('.input-wallet')
          .simulate('change', { target: { value: 'abc' } });
        wallet.find('.btn-deposit').simulate('click');
        expect(mockDeposit).toHaveBeenCalledTimes(0);
        wallet
          .find('.input-wallet')
          .simulate('change', { target: { value: 'afsfa.f84safa' } });
        wallet.find('.btn-deposit').simulate('click');
        expect(mockDeposit).toHaveBeenCalledTimes(0);
      });
    });

    describe('and the user wants to make a withdrawal', () => {
      beforeEach(() => {
        wallet.find('.btn-withdraw').simulate('click');
      });

      it('dispatches the `withdraw()` it recieves from props with the local balance', () => {
        expect(mockWithrawal).toHaveBeenCalledWith(parseInt(userBalance, 10));
      });

      it("doesn't dispatch the `withdraw()` if a non number is recieved from props with local balance", () => {
        wallet
          .find('.input-wallet')
          .simulate('change', { target: { value: 'abc' } });
        wallet.find('.btn-withdraw').simulate('click');
        expect(mockDeposit).toHaveBeenCalledTimes(0);
        wallet
          .find('.input-wallet')
          .simulate('change', { target: { value: 'afsfa.f84safa' } });
        wallet.find('.btn-withdraw').simulate('click');
        expect(mockDeposit).toHaveBeenCalledTimes(0);
      });
    });
  });
});
