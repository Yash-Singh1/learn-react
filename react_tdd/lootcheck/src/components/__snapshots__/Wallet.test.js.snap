// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet renders properly 1`] = `
Wallet {
  "context": Object {},
  "props": Object {
    "balance": 20,
    "deposit": [MockFunction],
    "withdraw": [MockFunction],
  },
  "refs": Object {},
  "state": Object {
    "balance": undefined,
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useDeferredValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useResponder": [Function],
        "useState": [Function],
        "useTransition": [Function],
      },
      "_element": <Wallet
        balance={20}
        deposit={[MockFunction]}
        withdraw={[MockFunction]}
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <div>
        <h3
          className="balance"
        >
          Wallet balance: 
          20
        </h3>
        <br />
        <input
          className="input-wallet"
          onChange={[Function]}
        />
        <Button
          active={false}
          className="btn-deposit"
          disabled={false}
          onClick={[Function]}
          variant="primary"
        >
          Deposit
        </Button>
        <Button
          active={false}
          className="btn-withdraw"
          disabled={false}
          onClick={[Function]}
          variant="primary"
        >
          Withdraw
        </Button>
      </div>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;
